perf/src/recycler.rs

```diff
@@ -289,63 +88,69 @@ impl<T: Default + Reset + Sized> Recycler<T> {
+            //      a <- a * (n - 1) / n + x / n
+            // To avoid floating point math, define b = n a:
+            //      b <- b * (n - 1) / n + x
+            // To make the remaining division to round (instead of truncate),
+            // add n/2 to the numerator.
+            // Effectively b (size_factor here) is an exponential moving
+            // estimate of the "sum" of x (gc.len()) over the window as opposed
+            // to the "average".
+            self.recycler.size_factor.store(
+                self.recycler
+                    .size_factor
+                    .load(Ordering::Acquire)
+                    .saturating_mul(RECYCLER_SHRINK_WINDOW_SUB_ONE)
+                    .saturating_add(RECYCLER_SHRINK_WINDOW_HALF)
+                    .checked_div(RECYCLER_SHRINK_WINDOW)
+                    .unwrap()
+                    .saturating_add(gc.len()),
+                Ordering::Release,
+            );
+            if let Some(mut x) = gc.pop() {
                 self.recycler.stats.reuse.fetch_add(1, Ordering::Relaxed);
-                reused_object.reset();
-                Some(reused_object)
+                x.reset();
+                return x;
             }
-            AllocationDecision::Allocate(total_allocated_count, recycled_len) => {
-                let mut t = T::default();
-                t.set_recycler(Arc::downgrade(&self.recycler));
-                if total_allocated_count % 1000 == 0 {
-                    datapoint_info!(
-                        "recycler_total_allocated_count",
-                        ("name", self.shrink_metric_name, String),
-                        ("count", total_allocated_count as i64, i64),
-                        ("recycled_len", recycled_len as i64, i64),
-                    )
-                }
-                Some(t)
-            }
-
-            AllocationDecision::AllocationLimitReached => None,
         }
-    }
+        let total = self.recycler.stats.total.fetch_add(1, Ordering::Relaxed);
+        trace!(
+            "allocating new: total {} {:?} id: {} reuse: {} max_gc: {}",
+            total,
+            name,
+            self.recycler.id,
+            self.recycler.stats.reuse.load(Ordering::Relaxed),
+            self.recycler.stats.max_gc.load(Ordering::Relaxed),
+        );
 
-    pub fn recycle_for_test(&self, x: T) {
-        self.recycler.recycle(x);
+        let mut t = T::default();
+        t.set_recycler(Arc::downgrade(&self.recycler));
+        t
     }
 }
 
 impl<T: Default + Reset> RecyclerX<T> {
     pub fn recycle(&self, x: T) {
         let len = {
             let mut gc = self.gc.lock().expect("recycler lock in pub fn recycle");
-            gc.objects.push(x);
+            gc.push(x);
+            const SIZE_FACTOR_AFTER_SHRINK: usize = RECYCLER_SHRINK_SIZE * RECYCLER_SHRINK_WINDOW;
+            if gc.len() > RECYCLER_SHRINK_SIZE
+                && self.size_factor.load(Ordering::Acquire) >= SIZE_FACTOR_AFTER_SHRINK
```

Comment:
Hi! I'm called.  :) the exponential stuff looks like really cool. I'm still grokking the code... I'm weak at math to be honest...

That being said, I think this is maybe susceptible to eclipse attack? So, I think we need some highwater mark check to force shrinking if `gc.len()` is too large? If I read the code correctly, there is no upper bound for `gc.len()`?

Consider this: victim's network traffic shaping is patially under the control by malicious actor like network intermediaries or turbine neighbor (PinnedVec/Recycler is used on that codepath?) (I know this is a bit too theoretical...)

Attack: the malicious actor throttles packets with intermittent spikes to the victim so that `size_factor` is unintendedly below `SIZE_FACTOR_AFTER_SHRINK` although `gc.len()` is swollen because of processing the spike.


Also, I'm still wondering  new impl is as safe as older shrinker and limiter one. My gut feeling is that centralized hard-limit at lower layer is safer while I know it contained bugs that frustrated @behzadnouri due to my lack of review at #15320... I'm still trying convincing this cleaner impl is actually ok. :)

Anyway, let's have a happy discussion. the more the merrier.

Reply:
The goal is to pull rate-limiting and allocation limit out of the recycler to higher level code which has more context. So,
* The recycler will just do the recycling and shrinking its internal buffer when necessary (i.e. this pull request); which
    * Allows simpler code, as the recycler will not bother with tracking number of objects on the fly.
    * Avoids several bugs with current implementation as described in the pull-request description.

and, then follow up with:
* Use things like bounded channels to enforce rate limits.
* Use things like load-shedding to limit memory usage; as in for example in gossip:
  https://github.com/solana-labs/solana/blob/364af3a3e/core/src/cluster_info.rs#L2811-L2822

Reply:
@behzadnouri 

I like your exponential shrinker impl in fact. it's just beautiful.

That being said, let's focus on whether or not we should rare-limit at lower layer. :)

I feel the merit of your idea of moving these concerns (rate-limit, allocation limit) to higher layer. But, that would impose all of higher level individual components must be equipped with rate-limits and allocation limits, maybe?

 That sounds tedious to me. Also, I'm thinking ideally we want to introduce some flow-control somewhere (preferably, at the common lower layer) in the future (@leoluk I'm not forgetting you... https://github.com/solana-labs/solana/issues/12410). So, I'm slightly inclined to make the lower layer smarter not dumber.

Similar to the general data sanitization strategy in spirit, I'd rather to have these limits at lower layer _if lack of higher level context doesn't introduce too much trouble_. I think this is yes to some extent for now.

@behzadnouri @carllin Could you share you opinion on this after listening to me? Thanks in advance. :)

Firstly, solana-validator is kind of (insanely-fast) steaming-processor at the highest level in this packet processing aspect. It's constantly fed with saturated 1G or 10G incoming bandwidth. And those bytes are untrusted, so must be cryptographically-verified. So, we want to fed the opaque blobs/bytes to GPU for that end as smooth as possible before any further processing.

What I want to say here is that:
1. The lower layer adequately knows the upper bound of input rate. Even it can be aware of the nature of shared bandwidth between other incoming packet sources
1. Higher layer does create higher-level `struct` from bytes but won't create more than what's coming from the wire (unless we have some data amplification bug). That means I think the higher layer isn't necessarily smarter than lower layer. (But the higher layer must adjust their `struct`s lifetimes not to pressure mem by pruning older data in their definition)
1. we don't care about out-going packets as for rate-limiting. In other words, higher layer doesn't create PinnedVec out of nowhere internally at undetermined rate; PinnedVec is always tied to the incoming network link bandwidth.
1. we need some tighter control over the mem region for gpu pinning.

And here's `PinnedVec` and `Recyler` toys for that end, which tries to reuse same buffer over and over for better data-locality and to avoid allocation cost and reduce calling potentially costly pinning api call for Cuda.

From this high-level design philosophy, I think it's actually bad coding pattern for `clone()`-ing or _consuming_ off `PinnedVecs`. Doing these isn't efficient and it also crates another risk of potential mem bloating hidden _buffer_. But the current code allows that, confusing its motto. Also, its implementation part actually had bugs as @behzadnouri has pointed with this pr. 

So, here's my draft patch to remove these PinnedVec blob escapes, so fixing the bugs from another different angle: https://github.com/ryoqun/solana/tree/revert-alloc-limit-no-clone

At the patch, I'm making sure `PinnedVec` is something which should only be *referenced*, not moved or cloned. So, `Recycler` can clearer understanding of current allocation situation for possible smarter decision in the future code (if we decide to do so).

I wanted to remove `.clones()` altogether to begin with at https://github.com/solana-labs/solana/pull/15320#issuecomment-781296765. but blame me, no time... ;)
Also, note that the patch is in draft: it only complies with `cargo clippy --bin solana-validator`. lol. I'm just showing that my idea is possible with the current codebase.

So fun to have these engineering discussions among the team. :)

Reply:
Also this is a handy diff view for my branch: https://github.com/solana-labs/solana/compare/99d12b8...ryoqun:revert-alloc-limit-no-clone?expand=1

Reply:
@sakridge by the way, the above comment makes sense as the oldest member who wrote the relevant code? That's what I sensed only from reading the code without any briefing. lol

Also, as for coalscing packets for banking stage, it might make sense to create thin struct like `BankPacket[s]` or `TxPacket[s]` (bad naming) to follow the higher-layer specific `struct`s design.

Anyway, this is my 2 cents.

Reply:
I disagree.

I think recycler just better to do the recycling, and leave the rate-limiting, allocation limits, load shedding, etc to higher level code which has more context, and can do so *explicitly*.

Putting the allocation limit in the recycler is a fundamentally wrong design to me, leading to very hard to debug bugs as the ones which wasted me several days.

and, please note that you were not aware of these bugs. I was the one who was finding these bugs, and there may be even more corner cases around.

Reply:
@ryoqun Yea, I think your comment makes sense in terms of the reasons for `PinnedVec` and `Recycler` and how they are designed to keep allocations around for re-use.

I would say at the time rust has no support for custom allocators for `Vec` so the semantics around `PinnedVec` are a bit different from the normal rust semantics. One has to be careful to consume the memory in the right way to get it recycled and mapped/unmapped in the right way unlike a normal `Vec` with fungible memory space.

I agree with @behzadnouri these probably should be at a higher level since recycler may not have the information for shedding and rate limiting for all the protocols that it may use. The best pruning/rate limiting algorithms may be protocol specific in terms of what packets you want to filter and prioritize, so recycler is not the right place even if it is convenient.

Reply:
@behzadnouri thanks for additional points. Okay. point taken. let's remove the limit from recycler.

If I regard the recycler like a  system-provided malloc (which basically should never fail), I can understand it's odd place to add limit as you say. Previously, I was thinking it's like an io scheduler/network shaping queue or something backed by shared common resource (with limited throughput).

Anyway, our goal here is just to build very robust and fast stream processor under completely byzantine environment _as fast as possible_. I'm looking forward to seeing you to implement proper limiters on respective higher-level protocols in follow-up prs. :)


@sakridge Thanks for chiming in! Glad to hear my understanding isn't far off at least. :)